aptos-move/parallel-executor/src/txn_last_input_output.rs

```diff
@@ -76,17 +86,58 @@ impl<K, T: TransactionOutput, E: Send + Clone> TxnLastInputOutput<K, T, E> {
             outputs: (0..num_txns)
                 .map(|_| CachePadded::new(ArcSwapOption::empty()))
                 .collect(),
+            written_module_paths: DashSet::new(),
+            read_module_paths: DashSet::new(),
+        }
+    }
+
+    fn append_module_paths(
+        paths: Vec<AccessPath>,
+        set_to_append: &DashSet<AccessPath>,
+        set_to_check: &DashSet<AccessPath>,
+    ) -> Result<()> {
+        for path in paths {
+            // Standard flags, first show, then look.
+            set_to_append.insert(path.clone());
+
+            if set_to_check.contains(&path) {
+                bail!("Module published and also read");
+            }
         }
+        Ok(())
     }
 
     pub fn record(
         &self,
         txn_idx: TxnIndex,
         input: Vec<ReadDescriptor<K>>,
         output: ExecutionStatus<T, Error<E>>,
-    ) {
+    ) -> Result<()> {
+        let read_modules: Vec<AccessPath> =
+            input.iter().filter_map(|desc| desc.module_path()).collect();
```

Comment:
I don't think this is correct since vm doesn't expose cache hits? we'll probably need to leverage @wrwg's new change

Reply:
The first miss (when it's loaded) will be recorded for sure because otherwise nothing can make it into the VM loader cache. Since we put everything into the same Dashset, we should never miss any module that's used.

Reply:
Wolfgang mentioned this invariant as well, I'll definitely add a comment to explain the reasoning.

Reply:
oh it's per whole block, so I guess it's correct but just confusing

Reply:
added comments.